File: src/main/java/com/kaushik/restapis/bookstore_management/util/DateUtils.java
package com.kaushik.restapis.bookstore_management.util;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Utility class for date and time operations. Provides common date formatting
 * and manipulation methods.
 *
 * @author Teammate Name
 * @version 1.0
 */
public class DateUtils {

    private static final DateTimeFormatter DEFAULT_DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd");
    private static final DateTimeFormatter DEFAULT_DATETIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
    private static final DateTimeFormatter DISPLAY_DATE_FORMATTER = DateTimeFormatter.ofPattern("MMM dd, yyyy");
    private static final DateTimeFormatter DISPLAY_DATETIME_FORMATTER = DateTimeFormatter.ofPattern("MMM dd, yyyy HH:mm:ss");

    private DateUtils() {
        // Private constructor to prevent instantiation
        throw new UnsupportedOperationException("Utility class cannot be instantiated");
    }

    /**
     * Formats a LocalDate to default format (yyyy-MM-dd).
     *
     * @param date the date to format
     * @return formatted date string
     */
    public static String formatDate(LocalDate date) {
        if (date == null) {
            return null;
        }
        return date.format(DEFAULT_DATE_FORMATTER);
    }

    /**
     * Formats a LocalDateTime to default format (yyyy-MM-dd HH:mm:ss).
     *
     * @param dateTime the datetime to format
     * @return formatted datetime string
     */
    public static String formatDateTime(LocalDateTime dateTime) {
        if (dateTime == null) {
            return null;
        }
        return dateTime.format(DEFAULT_DATETIME_FORMATTER);
    }

    /**
     * Formats a LocalDate to display format (MMM dd, yyyy). Example: Nov 03,
     * 2025
     *
     * @param date the date to format
     * @return formatted date string for display
     */
    public static String formatDateForDisplay(LocalDate date) {
        if (date == null) {
            return null;
        }
        return date.format(DISPLAY_DATE_FORMATTER);
    }

    /**
     * Formats a LocalDateTime to display format (MMM dd, yyyy HH:mm:ss).
     * Example: Nov 03, 2025 10:30:45
     *
     * @param dateTime the datetime to format
     * @return formatted datetime string for display
     */
    public static String formatDateTimeForDisplay(LocalDateTime dateTime) {
        if (dateTime == null) {
            return null;
        }
        return dateTime.format(DISPLAY_DATETIME_FORMATTER);
    }

    /**
     * Parses a date string in default format (yyyy-MM-dd).
     *
     * @param dateStr the date string to parse
     * @return LocalDate object
     * @throws java.time.format.DateTimeParseException if the string cannot be
     * parsed
     */
    public static LocalDate parseDate(String dateStr) {
        if (dateStr == null || dateStr.trim().isEmpty()) {
            return null;
        }
        return LocalDate.parse(dateStr, DEFAULT_DATE_FORMATTER);
    }

    /**
     * Parses a datetime string in default format (yyyy-MM-dd HH:mm:ss).
     *
     * @param dateTimeStr the datetime string to parse
     * @return LocalDateTime object
     * @throws java.time.format.DateTimeParseException if the string cannot be
     * parsed
     */
    public static LocalDateTime parseDateTime(String dateTimeStr) {
        if (dateTimeStr == null || dateTimeStr.trim().isEmpty()) {
            return null;
        }
        return LocalDateTime.parse(dateTimeStr, DEFAULT_DATETIME_FORMATTER);
    }

    /**
     * Checks if a date is in the past.
     *
     * @param date the date to check
     * @return true if date is before today, false otherwise
     */
    public static boolean isPast(LocalDate date) {
        if (date == null) {
            return false;
        }
        return date.isBefore(LocalDate.now());
    }

    /**
     * Checks if a date is in the future.
     *
     * @param date the date to check
     * @return true if date is after today, false otherwise
     */
    public static boolean isFuture(LocalDate date) {
        if (date == null) {
            return false;
        }
        return date.isAfter(LocalDate.now());
    }

    /**
     * Checks if a date is today.
     *
     * @param date the date to check
     * @return true if date is today, false otherwise
     */
    public static boolean isToday(LocalDate date) {
        if (date == null) {
            return false;
        }
        return date.isEqual(LocalDate.now());
    }

    /**
     * Gets the current date as a formatted string.
     *
     * @return current date in default format
     */
    public static String getCurrentDate() {
        return formatDate(LocalDate.now());
    }

    /**
     * Gets the current datetime as a formatted string.
     *
     * @return current datetime in default format
     */
    public static String getCurrentDateTime() {
        return formatDateTime(LocalDateTime.now());
    }

    /**
     * Calculates the number of days between two dates.
     *
     * @param startDate the start date
     * @param endDate the end date
     * @return number of days between the dates
     */
    public static long daysBetween(LocalDate startDate, LocalDate endDate) {
        if (startDate == null || endDate == null) {
            return 0;
        }
        return java.time.temporal.ChronoUnit.DAYS.between(startDate, endDate);
    }

    /**
     * Adds days to a date.
     *
     * @param date the base date
     * @param days number of days to add (can be negative)
     * @return new date with days added
     */
    public static LocalDate addDays(LocalDate date, long days) {
        if (date == null) {
            return null;
        }
        return date.plusDays(days);
    }

    /**
     * Adds months to a date.
     *
     * @param date the base date
     * @param months number of months to add (can be negative)
     * @return new date with months added
     */
    public static LocalDate addMonths(LocalDate date, long months) {
        if (date == null) {
            return null;
        }
        return date.plusMonths(months);
    }

    /**
     * Adds years to a date.
     *
     * @param date the base date
     * @param years number of years to add (can be negative)
     * @return new date with years added
     */
    public static LocalDate addYears(LocalDate date, long years) {
        if (date == null) {
            return null;
        }
        return date.plusYears(years);
    }

    /**
     * Gets the start of day for a given date.
     *
     * @param date the date
     * @return LocalDateTime at 00:00:00
     */
    public static LocalDateTime startOfDay(LocalDate date) {
        if (date == null) {
            return null;
        }
        return date.atStartOfDay();
    }

    /**
     * Gets the end of day for a given date.
     *
     * @param date the date
     * @return LocalDateTime at 23:59:59
     */
    public static LocalDateTime endOfDay(LocalDate date) {
        if (date == null) {
            return null;
        }
        return date.atTime(23, 59, 59);
    }
}

File: frontend/src/shared/utils/dateUtils.ts
// Utility functions for date operations
import {
  format,
  parse,
  isValid,
  differenceInDays,
  addDays,
  addMonths,
  addYears,
} from "date-fns";

/**
 * Formats a date to a specified format
 */
export const formatDate = (
  date: Date | string,
  formatStr: string = "MMM dd, yyyy"
): string => {
  try {
    const dateObj = typeof date === "string" ? new Date(date) : date;
    if (!isValid(dateObj)) return "";
    return format(dateObj, formatStr);
  } catch {
    return "";
  }
};

/**
 * Formats a date for display (e.g., "Nov 03, 2025")
 */
export const formatDateForDisplay = (date: Date | string): string => {
  return formatDate(date, "MMM dd, yyyy");
};

/**
 * Formats a date for API (ISO format)
 */
export const formatDateForAPI = (date: Date | string): string => {
  try {
    const dateObj = typeof date === "string" ? new Date(date) : date;
    if (!isValid(dateObj)) return "";
    return dateObj.toISOString().split("T")[0];
  } catch {
    return "";
  }
};

/**
 * Formats a datetime for display
 */
export const formatDateTime = (
  date: Date | string,
  formatStr: string = "MMM dd, yyyy HH:mm"
): string => {
  return formatDate(date, formatStr);
};

/**
 * Parses a date string
 */
export const parseDate = (
  dateStr: string,
  formatStr: string = "yyyy-MM-dd"
): Date | null => {
  try {
    const parsed = parse(dateStr, formatStr, new Date());
    return isValid(parsed) ? parsed : null;
  } catch {
    return null;
  }
};

/**
 * Checks if a date is in the past
 */
export const isPast = (date: Date | string): boolean => {
  const dateObj = typeof date === "string" ? new Date(date) : date;
  return dateObj < new Date();
};

/**
 * Checks if a date is in the future
 */
export const isFuture = (date: Date | string): boolean => {
  const dateObj = typeof date === "string" ? new Date(date) : date;
  return dateObj > new Date();
};

/**
 * Checks if a date is today
 */
export const isToday = (date: Date | string): boolean => {
  const dateObj = typeof date === "string" ? new Date(date) : date;
  const today = new Date();
  return (
    dateObj.getDate() === today.getDate() &&
    dateObj.getMonth() === today.getMonth() &&
    dateObj.getFullYear() === today.getFullYear()
  );
};

/**
 * Gets the number of days between two dates
 */
export const daysBetween = (
  date1: Date | string,
  date2: Date | string
): number => {
  const d1 = typeof date1 === "string" ? new Date(date1) : date1;
  const d2 = typeof date2 === "string" ? new Date(date2) : date2;
  return differenceInDays(d2, d1);
};

/**
 * Adds days to a date
 */
export const addDaysToDate = (date: Date | string, days: number): Date => {
  const dateObj = typeof date === "string" ? new Date(date) : date;
  return addDays(dateObj, days);
};

/**
 * Adds months to a date
 */
export const addMonthsToDate = (date: Date | string, months: number): Date => {
  const dateObj = typeof date === "string" ? new Date(date) : date;
  return addMonths(dateObj, months);
};

/**
 * Adds years to a date
 */
export const addYearsToDate = (date: Date | string, years: number): Date => {
  const dateObj = typeof date === "string" ? new Date(date) : date;
  return addYears(dateObj, years);
};

/**
 * Gets a relative time string (e.g., "2 days ago", "in 3 hours")
 */
export const getRelativeTime = (date: Date | string): string => {
  const dateObj = typeof date === "string" ? new Date(date) : date;
  const now = new Date();
  const seconds = Math.floor((now.getTime() - dateObj.getTime()) / 1000);

  if (seconds < 60) return "just now";
  if (seconds < 3600) return `${Math.floor(seconds / 60)} minutes ago`;
  if (seconds < 86400) return `${Math.floor(seconds / 3600)} hours ago`;
  if (seconds < 604800) return `${Math.floor(seconds / 86400)} days ago`;
  if (seconds < 2592000) return `${Math.floor(seconds / 604800)} weeks ago`;
  if (seconds < 31536000) return `${Math.floor(seconds / 2592000)} months ago`;
  return `${Math.floor(seconds / 31536000)} years ago`;
};

/**
 * Gets the start of day
 */
export const startOfDay = (date: Date | string): Date => {
  const dateObj = typeof date === "string" ? new Date(date) : date;
  const result = new Date(dateObj);
  result.setHours(0, 0, 0, 0);
  return result;
};

/**
 * Gets the end of day
 */
export const endOfDay = (date: Date | string): Date => {
  const dateObj = typeof date === "string" ? new Date(date) : date;
  const result = new Date(dateObj);
  result.setHours(23, 59, 59, 999);
  return result;
};

/**
 * Validates if a date string is valid
 */
export const isValidDate = (dateStr: string): boolean => {
  const date = new Date(dateStr);
  return isValid(date);
};

/**
 * Gets the current date as a string
 */
export const getCurrentDate = (): string => {
  return formatDateForAPI(new Date());
};

/**
 * Gets the current datetime as a string
 */
export const getCurrentDateTime = (): string => {
  return new Date().toISOString();
};
